frontend/catalyst/autograph.py

```diff
@@ -0,0 +1,158 @@
+        At a minimum we need to provide the module with definitions for Autograph primitives."""
+
+        return {"ag__": ag_primitives}
+
+    def get_caching_key(self, user_context):
+        """Autograph automatically caches transformed functions, the caching key is a combination of
+        the function source as well as a custom key provided by us here. Changing Autograph options
+        should trigger the function transform again, rather than getting it from cache."""
+
+        return user_context.options
+
+    def has_cache(self, fn, cache_key):
+        """Check for the presence of the given function in the cache. Functions to be converted are
+        cached by the function object itself as well as the conversion options."""
+
+        return self._cache.has(fn, cache_key)
+
+    def get_cached_function(self, fn, cache_key):
+        """Retrieve a Python function object for a previously converted function.
+        Note that repeatedly calling this function with the same arguments will result in new
+        function objects every time, however their source code should be identical with the
+        exception of auto-generated names."""
+
+        # Converted functions are cached as a _PythonFnFactory object.
+        cached_factory = self._cached_factory(fn, cache_key)
+
+        # Convert to a Python function object before returning (e.g. to obtain its source code).
+        new_fn = cached_factory.instantiate(
+            fn.__globals__,
+            fn.__closure__ or (),
+            defaults=fn.__defaults__,
+            kwdefaults=getattr(fn, "__kwdefaults__", None),
+        )
+
+        return new_fn
+
+
+def autograph(fn):
+    """Decorator that converts the given function into graph form."""
+
+    options = converter.ConversionOptions(
+        recursive=True,
+        user_requested=True,
+        internal_convert_user_code=True,
+        optional_features=None,
+    )
+    user_context = converter.ProgramContext(options)
+
+    new_fn, module, source_map = _TRANSFORMER.transform(fn, user_context)
+    new_fn.ag_module = module
+    new_fn.ag_source_map = source_map
+    new_fn.ag_unconverted = fn
+
+    return new_fn
+
+
+def print_code(fn):
+    """Utility function to retrieve the source code of a converted function, in particular of
+    functions that were recursively converted. Note that recursive conversion is only triggered
+    *after* the @autograph decorated function is called at least once."""
+
+    cache_key = ag_primitives.STD
+
+    if hasattr(fn, "ag_unconverted"):
+        # This is a function directly decorated with @autograph or @qjit(autograph=True).
+        print(inspect.getsource(fn))
+
+    elif _TRANSFORMER.has_cache(fn, cache_key):
+        # This is a recursively converted function.
+        new_fn = _TRANSFORMER.get_cached_function(fn, cache_key)
+        print(inspect.getsource(new_fn))
+
+    else:
+        print("Could not print source code: Function has not been converted.")
```

Comment:
Should it be an exception instead?

Reply:
This could be one option, the other is to just print the source code anyways thus making the function more general.

Reply:
I would vote for the explicit behavior. Or it could be a `raise_on_error` flag. Callers can always use try-catch if the want.

Reply:
@josh146 What do you think?

I'm thinking of exposing a function `print_code` that would make it easy for users to print the source code of an autograph-transformed function (and I use it in tests). This is especially important for nested functions, because the transformed object is not available in the user namespace.

The question is, what should happen if the function `print_code` is applied to a function that has not been transformed?
- Should it print the source anyways so that it works on both transformed and non-transformed functions alike?
- Should it raise an error?
- Should it print the source but with a note or warning stating this is the source code of a non-transformed function?

Reply:
> What would it print for a non-transformed function? just the captured version given compile-time determined branches?

It would just print the raw source code that the user wrote, this is pre-tracing.

> It would potentially be a more useful function if it returned the string directly, rather than hardcoding in a print statement E.g., transformed_code(fn). This is because someone debugging might want to capture/manipulate the string in other ways. Another example could be logging; this way, someone could do logging.log(transformed_code(fn), logging.INFO)

That makes sense, I was just tired of writing the print function everywhere 😅 (I miss the print statement :/)

> Depends on the stated function of the, uh, function. If it is for autograph debugging, I would err on the side of yes. Of course, you could always make this an argument, e.g.,

I would say yes, it is primarily for debugging (or for the curious). Well really I wrote it for testing, but including it in the public API would primarily be for debugging.

> If we want this to be more visible/discoverable, I recommend:
>
> Importing it into __init__ and adding it to __all__ so that it appears in the documentation.
> Giving it a richer docstring with examples.
> I think this makes a lot of sense, for two reasons:

Definitely, I just haven't added this feature in a public facing way yet :)

Reply:
There is a problem importing objects from the autograph module into the top-level `__init__`, because of the optional tensorflow dependency there is no guarantee that these objects can be defined 🤔

Reply:
Added the debug function, but I still need to figure out how to handle the imports: https://github.com/PennyLaneAI/catalyst/pull/235/commits/63bb331acf7adde0496ef7b7758fe8d45c7dbc4a

Reply:
🤔

Reply:
I resolved the import issues here: https://github.com/PennyLaneAI/catalyst/pull/235/commits/51914cc5543c2dc15ee93f325fd3cf01b27b77d2